### ISSUE TITLE
junos_config with "replace: override" fails

### ISSUE THREAD

**Issue by desht** [ansible/ansible#54107]
##### SUMMARY
When trying to replace some config on a JUNOS firewall using the ``junos_config`` module and ``replace: override``, the playbook fails with a ``Missing mandatory statement: 'root-authentication'`` JUNOS error, even though the ``root-authentication`` statement is definitely not missing from the firewall configuration.  The same task with ``replace: update`` does not fail, but also does not remove config from the firewall which is not in the candidate configuration.

Note: full file paths, device names, usernames, etc. are redacted from various output below.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
junos_config

##### ANSIBLE VERSION

```
ansible 2.7.9
  config file = /***/ansible.cfg
  configured module search path = [u'/usr/share/ansible', u'/etc/ansible/roles/Juniper.junos/library']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible-2.7.9-py2.7.egg/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Nov  1 2018, 03:12:47) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36.0.1)]
```

##### CONFIGURATION

```
ANSIBLE_SSH_RETRIES(***/ansible.cfg) = 2
DEFAULT_CALLBACK_WHITELIST(***/ansible.cfg) = [u'auditlog2']
DEFAULT_FORKS(***/ansible.cfg) = 20
DEFAULT_HOST_LIST(***/ansible.cfg) = [u'***/inventory']
DEFAULT_LOG_PATH(***/ansible.cfg) = ***/ansible.log
DEFAULT_MODULE_PATH(***/ansible.cfg) = [u'/usr/share/ansible', u'/etc/ansible/roles/Juniper.junos/library'
DEFAULT_ROLES_PATH(***/ansible.cfg) = [u'***/ansible/roles']
DEFAULT_STDOUT_CALLBACK(***/ansible.cfg) = skippy
DEFAULT_TIMEOUT(***/ansible.cfg) = 60
HOST_KEY_CHECKING(***/ansible.cfg) = False
PERSISTENT_COMMAND_TIMEOUT(***/ansible.cfg) = 300
PERSISTENT_CONNECT_TIMEOUT(***/ansible.cfg) = 300


```

##### OS / ENVIRONMENT

```
$ uname -a
Linux base-latest 4.1.12-124.19.2.el7uek.x86_64 #2 SMP Fri Sep 14 08:59:15 PDT 2018 x86_64 x86_64 x86_64 GNU/Linux
```

##### STEPS TO REPRODUCE

Task:
```
- name: Configure SNMP on a JUNOS SRX/MX device
  junos_config:
    provider: "...."
    src: junos-snmp.conf.j2
    src_format: text
    # update: replace
    update: override
  register: config_changes
```
Template:
```
snmp {
{% for item in snmp_query_all %}
  client-list "{{ item.community }}-clients" {
{% for client in item.clients %}
    {{ client }};
{% endfor %}
  }
  community "{{ item.community }}" {
    authorization {{ item.auth_level }};
    client-list-name "{{ item.community }}-clients";
  }
{% endfor %}
}
```

##### EXPECTED RESULTS

With ``update: override`` I would expect any SNMP configuration on the device to be overridden by the candidate config, removing any SNMP communities on the device which are not in the candidate config, and adding/overriding any communities which are in the candidate config.

##### ACTUAL RESULTS

```
The full traceback is:
Traceback (most recent call last):
  File "/home/user/.ansible/tmp/ansible-tmp-1553090101.53-230335757920271/AnsiballZ_junos_config.py", line 113, in <module>
    _ansiballz_main()
  File "/home/user/.ansible/tmp/ansible-tmp-1553090101.53-230335757920271/AnsiballZ_junos_config.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/user/.ansible/tmp/ansible-tmp-1553090101.53-230335757920271/AnsiballZ_junos_config.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_junos_config_payload_eH_hgL/__main__.py", line 405, in <module>
  File "/tmp/ansible_junos_config_payload_eH_hgL/__main__.py", line 373, in main
  File "/tmp/ansible_junos_config_payload_eH_hgL/__main__.py", line 295, in configure_device
  File "/tmp/ansible_junos_config_payload_eH_hgL/ansible_junos_config_payload.zip/ansible/module_utils/network/junos/junos.py", line 251, in load_config
  File "/tmp/ansible_junos_config_payload_eH_hgL/ansible_junos_config_payload.zip/ansible/module_utils/network/common/netconf.py", line 76, in __rpc__
  File "/tmp/ansible_junos_config_payload_eH_hgL/ansible_junos_config_payload.zip/ansible/module_utils/network/common/netconf.py", line 108, in parse_rpc_error
ansible.module_utils.connection.ConnectionError: error: mgd: Missing mandatory statement: 'root-authentication'
error: configuration check-out failed: (missing mandatory statements)

fatal: [devicename]: FAILED! => {
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/home/user/.ansible/tmp/ansible-tmp-1553090101.53-2303357
57920271/AnsiballZ_junos_config.py\", line 113, in <module>\n    _ansiballz_main()\n  File \"/home/user/.ansible/tmp/ansible-tmp-1553090101.53-230335757920271/AnsiballZ_junos_config.py\", line 105, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/user/.ansible/tmp/ansible-tmp-1553090101.53-230335757920271/AnsiballZ_junos_config.py\", line 48, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_junos_config_payload_eH_hgL/__main__.py\", line 405, in <module>\n  File \"/tmp/ansible_junos_config_payload_eH_hgL/__main__.py\", line 373, in main\n  File \"/tmp/ansible_junos_config_payload_eH_hgL/__main__.py\", line 295, in configure_device\n  File \"/tmp/ansible_junos_config_payload_eH_hgL/ansible_junos_config_payload.zip/ansible/module_utils/network/junos/junos.py\", line 251, in load_config\n  File \"/tmp/ansible_junos_config_payload_eH_hgL/ansible_junos_config_payload.zip/ansible/module_utils/network/common/netconf.py\", line 76, in __rpc__\n  File \"/tmp/ansible_junos_config_payload_eH_hgL/ansible_junos_config_payload.zip/ansible/module_utils/network/common/netconf.py\", line 108, in parse_rpc_error\nansible.module_utils.connection.ConnectionError: error: mgd: Missing mandatory statement: 'root-authentication'\nerror: configuration check-out failed: (missing mandatory statements)\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", 
    "rc": 1
}
```

This suggests that the device is missing the ``root-authentication`` statement, a common problem when trying to configure new JUNOS devices.

However, this device is not new, and is most certainly not missing the ``root-authentication`` statement.  In addition, I can modify and commit config without error when logged into the device on the command-line, and running the same task with ``update: replace`` does not fail (but also does not update the config as necessary - new config is added but unwanted config is not removed).  The above failure only occurs when using ``junos_config`` with ``update: override``.

I noticed also https://github.com/ansible/ansible/issues/40267 which implies some work was done with this in the past and cherry-picked from 2.8 into 2.7.6, but I'm running 2.7.9 here.  Not sure if this problem is related.

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/network/junos/junos_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/junos/junos_config.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Qalthos @ganeshrn @privateip
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@desht, just so you are aware we have a dedicated Working Group for network.
You can find other people interested in this in `#ansible-network` on Freenode IRC
For more information about communities, meetings and agendas see https://github.com/ansible/community

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ganeshrn** [comment]
`update: override` option tries to replace the entire configuration with the input configuration. Since `root-authentication` related config is not present in your input configuration file it results in an error. 

For the overirde case, the input config to the task should reflect the entire final state running configuration.

**Comment by desht** [comment]
Thanks, that makes sense.  I had mistakenly understood it to just override the config section I was trying to upload.

However, that still leaves a problem: removing unwanted entries.  For example, if my current SNMP config looks like:

```
snmp {
  client-list comm1-clients {
    1.2.3.4/32;
  }
  client-list comm2-clients {
    5.6.7.8/32
  }
  community comm1 {
    authorization read-only;
    client-list comm1-clients;
  }
  community comm2 {
    authorization read-only;
    client-list comm2-clients;
  }
}
```
and I apply this template:
```
snmp {
  client-list comm1-clients {
    1.2.3.4/32;
  }
  client-list comm3-clients {
    192.168.1.0/24;
  }
  community comm1 {
    authorization read-only;
    client-list comm1-clients;
  }
  community comm3 {
    authorization read-only;
    client-list comm3-clients;
  }
```
it will correctly add ``comm3`` but it won't remove ``comm2``.  This is a problem, since how I can remove an unwanted community from the device?  E.g. if we change our community strings (which we do periodically), I certainly don't want the old community string hanging around.

So I presume there's a way to have old config purged, and I was hoping ``update: replace`` would do that, but so far I haven't been able to see how.

**Comment by desht** [comment]
Actually, I think I've solved my own problem: the template requires a ``replace:`` tag right at the start:

```
replace:
snmp {
  .... rest of the template ...
}
```

I had experimented with ``replace:`` before but put it inside the ``snmp {`` section.  With hindsight, that was wrong, since ``replace:`` applies to the config that follows it.  In my case, I'm replacing the entire snmp section so this forces a replace rather than a merge.
